src/Raven.Server/Documents/Queries/QueryRunner.cs

```diff
@@ -92,7 +94,11 @@ public Task ExecutePatchQuery(string indexName, IndexQuery query, QueryOperation
             if (options.AllowStale == false && results.IsStale)
                 throw new InvalidOperationException("Cannot perform delete operation. Query is stale.");
 
-            foreach (var document in results.Results)
+            IEnumerable<Document> documents = results.Results;
+            if (options.MaxOpsPerSecond.HasValue)
+                documents = documents.LimitRate(options.MaxOpsPerSecond.Value, TimeSpan.FromSeconds(1));
+
+            foreach (var document in documents)
```

Comment:
Bad, you are holding a write transaction open with the rate limit lock taken.
So you effectively turned this into a _very_ long transaction.